components/MeasurementsChart.tsx

```diff
@@ -0,0 +1,249 @@
+import { Time, TimeSeries, timeSeries } from "pondjs";
+import { useCallback, useMemo } from "react";
+import {
+  CartesianGrid,
+  Label,
+  Line,
+  LineChart,
+  ReferenceDot,
+  ReferenceLine,
+  ResponsiveContainer,
+  Tooltip,
+  XAxis,
+  YAxis,
+} from "recharts";
+import { CategoricalChartState } from "recharts/types/chart/generateCategoricalChart";
+import { metricVarColor } from "../lib/api/models/metrics";
+import {
+  Measurement,
+  MetricName,
+  metricNames,
+  Timestamp,
+} from "../lib/api/types";
+import { formatDate } from "../lib/time";
+import { Legend } from "./Legend";
+import { MeasurementChartTooltip } from "./MeasurementChartTooltip";
+
+const halfADay = 1000 * 60 * 60 * 12;
+
+export type MeasurementsChartProps = {
+  data: Measurement[];
+  onDateSelected: (date: Timestamp) => void;
+  selectedDate?: Timestamp;
+};
+
+export const MeasurementsChart = ({
+  data,
+  onDateSelected,
+  selectedDate,
+}: MeasurementsChartProps) => {
+  const series = useMemo(() => {
+    const series = timeSeries({
+      name: "measurements",
+      columns: ["time", ...metricNames],
+      points: data.map((measurement) => {
+        return [
+          measurement.date,
+          ...metricNames.map((name) => measurement[name]),
+        ];
+      }),
+    });
+    return series;
+  }, [data]);
+
+  const onClick = useCallback(
+    (e: CategoricalChartState) => {
+      if (e && e.activePayload && e.activePayload.length > 0) {
+        const date = e.activePayload[0].payload?.date;
+        if (date) {
+          onDateSelected(date);
+        }
+      }
+    },
+    [onDateSelected]
+  );
+
+  const alertDots = useMemo(() => {
+    return data.flatMap((measurement) => {
+      return (measurement.notes || [])
+        .filter((note) => note.type === "alert")
+        .flatMap((note) => {
+          return note.metrics.map((metric) => {
+            return {
+              x: measurement.date + halfADay,
```

Comment:
Why are those moved by a half a day?
I think the graph behaves a bit strange with vertical grid lines and the lines that follows mouse pointer not being aligned with points displayed on graph.

I also think tooltip highlighting works a bit strange because of that as it always highlight points to the right of the mouse pointer.

Reply:
I agree with you Kamil.

It's not my idea. The root cause of this problem is what recharts library shows when `date` or even `linear` x-axis scale type is used:

<img width="350" alt="Zrzut ekranu 2022-04-4 o 13 25 40" src="https://user-images.githubusercontent.com/2854949/161534668-4ae3b35e-62f4-46fd-b164-1982e593f747.png">

A - line position aligned with actual x-axis value
B - dot is displayed in the middle of the day from some reason (even with `linear` scale)

I tried to fight with the library for an hour or so and left it as it is now, but I'm not satisfied with the result. I'd prefer to have line A and dot B in position A. It would need more time to find a solution. Maybe hiding dot B with css would solve this.

By the way, with default scale type it works as expected. The problem with default scale type is that the overall graph shape is distorted because gaps between days are not visible. I decided to show graph shape properly.